src/lib/Libutil/hook.c

```diff
@@ -3868,3 +3868,82 @@ cleanup_hooks_workdir(struct work_task *ptask)
 		cleanup_hooks_workdir, NULL);
 
 }
+
+/**
+ *
+ * @brief
+ * 	Start profiling the next lines of code, collectively giving it
+ *      a 'label' and an 'action' description.
+ *	The 'label' usually identifies to some object being profiled
+ *	(e.g. hook), while the 'action' describes what
+ *	is being captured for the object (e.g. "initialization").
+ *	Both 'label' and 'action' uniquely identify the
+ *	data collected.
+ *
+ * @param[in]	label - describes a particular object 
+ * @param[in]	action - refers to the object's action
+ *
+ * @param[in]	print_start_msg - if 1, then log a "profile_start" message. 
+ *
+ * @return void
+ *
+ */
+void
+hook_perf_stat_start(char *label, char *action, int print_start_msg)
+{
+	char instance[MAXBUFLEN];
+
+	if (!will_log_event(PBSEVENT_DEBUG4))
+		return;
```

Comment:
Is it possible for an admin to change the log event mask while a hook is running? That would mean hook_perf_stat_start() would not get called, but hook_perf_stat_stop() would. Might that cause problems?

Reply:
If hook_perf_stop() is called and yet hook_perf_start() was called but returned early since DEBUG4 was not in the log_event_mask at the time, then nothing would really happen since the 'instance' has not been added to the performance stats list; that is, hook_perf_stop() would not find an entry for 'instance', so nothing is logged. It's the corollary that needs to be addressed. If hook_perf_start() was called, and yet hook_perf_stop() returns early since it does not find DEBUG4 in log_event_mask, then we'll have a leftover instance entry, that is not freed, and may hang around forever. I've updated the code in this case to make sure to free up always the instance entry that was set up by hook_perf_start().